File: app/src/main/java/com/mrboomdev/awery/app/data/settings/SettingsActions.java
package com.mrboomdev.awery.app.data.settings;

import static com.mrboomdev.awery.app.App.showLoadingWindow;
import static com.mrboomdev.awery.app.App.toast;
import static com.mrboomdev.awery.app.Lifecycle.getAnyContext;
import static com.mrboomdev.awery.app.Lifecycle.startActivityForResult;
import static com.mrboomdev.awery.util.async.AsyncUtils.thread;
import static com.mrboomdev.awery.util.io.FileUtil.deleteFile;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.provider.Settings;
import android.util.Log;

import androidx.annotation.NonNull;

import com.mrboomdev.awery.R;
import com.mrboomdev.awery.app.CrashHandler;
import com.mrboomdev.awery.app.data.Constants;
import com.mrboomdev.awery.app.services.BackupService;
import com.mrboomdev.awery.app.update.UpdatesManager;
import com.mrboomdev.awery.ext.data.Setting;
import com.mrboomdev.awery.generated.AwerySettings;
import com.mrboomdev.awery.sdk.util.MimeTypes;
import com.mrboomdev.awery.ui.activity.AboutActivity;
import com.mrboomdev.awery.ui.activity.ExperimentsActivity;
import com.mrboomdev.awery.ui.activity.setup.SetupActivity;
import com.mrboomdev.awery.util.async.AsyncFuture;
import com.mrboomdev.awery.util.exceptions.CancelledException;
import com.mrboomdev.awery.util.exceptions.ExceptionDescriptor;

import org.jetbrains.annotations.Contract;

import java.io.File;
import java.util.Calendar;

import xcrash.XCrash;

@Deprecated(forRemoval = true)
public class SettingsActions {
	private static final String TAG = "SettingsActions";

	@Contract(pure = true)
	public static void run(@NonNull Setting item) {
		var actionName = item.getKey();
		if(actionName == null) return;

		switch(actionName) {
			case AwerySettings.CLEAR_IMAGE_CACHE -> {
				deleteFile(new File(getAnyContext().getCacheDir(), Constants.DIRECTORY_IMAGE_CACHE));
				toast(R.string.cleared_successfully);
			}

			case AwerySettings.SETUP_THEME -> {
				var context = getAnyContext();
				var intent = new Intent(context, SetupActivity.class);
				intent.putExtra(SetupActivity.EXTRA_STEP, SetupActivity.STEP_THEMING);
				intent.putExtra(SetupActivity.EXTRA_FINISH_ON_COMPLETE, true);
				context.startActivity(intent);
			}

			case AwerySettings.ABOUT -> {
				var context = getAnyContext();
				var intent = new Intent(context, AboutActivity.class);
				context.startActivity(intent);
			}

			case AwerySettings.CLEAR_WEBVIEW_CACHE -> {
				deleteFile(new File(getAnyContext().getCacheDir(), Constants.DIRECTORY_WEBVIEW_CACHE));
				toast(R.string.cleared_successfully);
			}

			case AwerySettings.CLEAR_NET_CACHE -> {
				deleteFile(new File(getAnyContext().getCacheDir(), Constants.DIRECTORY_NET_CACHE));
				toast(R.string.cleared_successfully);
			}

			case AwerySettings.BACKUP -> {
				var date = Calendar.getInstance();

				var defaultName = "awery_backup_[" + date.get(Calendar.YEAR) + "_" +
						date.get(Calendar.MONTH) + "_" +
						date.get(Calendar.DATE) + "]_[" +
						date.get(Calendar.HOUR_OF_DAY) + "_" +
						date.get(Calendar.MINUTE) + "].awerybck";

				var context = getAnyContext();
				var intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
				intent.addCategory(Intent.CATEGORY_OPENABLE);
				intent.setType(MimeTypes.ANY.toString());
				intent.putExtra(Intent.EXTRA_TITLE, defaultName);

				startActivityForResult(context, intent, 0, ((resultCode, data) -> {
					if(resultCode != Activity.RESULT_OK) return;

					var backupIntent = new Intent(context, BackupService.class);
					backupIntent.setAction(BackupService.ACTION_BACKUP);
					backupIntent.setData(data.getData());
					context.startService(backupIntent);
				}));
			}

			case AwerySettings.RESTORE -> {
				var context = getAnyContext();
				var intent = new Intent(Intent.ACTION_GET_CONTENT);
				intent.setType(MimeTypes.ANY.toString());
				var chooser = Intent.createChooser(intent, "Choose a backup file");

				startActivityForResult(context, chooser, 0, ((resultCode, data) -> {
					if(resultCode != Activity.RESULT_OK) return;

					var restoreIntent = new Intent(context, BackupService.class);
					restoreIntent.setAction(BackupService.ACTION_RESTORE);
					restoreIntent.setData(data.getData());
					context.startService(restoreIntent);
				}));
			}

			case AwerySettings.PLAYER_SYSTEM_SUBTITLES -> getAnyContext()
					.startActivity(new Intent(Settings.ACTION_CAPTIONING_SETTINGS));

			case AwerySettings.TRY_CRASH_NATIVE ->
					XCrash.testNativeCrash(false);

			case AwerySettings.TRY_CRASH_NATIVE_ASYNC ->
					thread(() -> XCrash.testNativeCrash(false));

			case AwerySettings.TRY_CRASH_JAVA ->
					XCrash.testJavaCrash(false);

			case AwerySettings.TRY_CRASH_JAVA_ASYNC ->
					thread(() -> XCrash.testJavaCrash(false));

			case AwerySettings.START_ONBOARDING -> {
				var context = getAnyContext();
				context.startActivity(new Intent(context, SetupActivity.class));
			}

			case AwerySettings.EXPERIMENTS -> {
				var context = getAnyContext();
				context.startActivity(new Intent(context, ExperimentsActivity.class));
			}

			case AwerySettings.CHECK_APP_UPDATE -> UpdatesManager.getAppUpdate().addCallback(new AsyncFuture.Callback<>() {
				private final Dialog window = showLoadingWindow();

				@Override
				public void onSuccess(UpdatesManager.Update update) {
					UpdatesManager.showUpdateDialog(update);
					window.dismiss();
				}

				@Override
				public void onFailure(Throwable t) {
					Log.e(TAG, "Failed to check for updates!", t);
					window.dismiss();

					if(t instanceof CancelledException) {
						toast(ExceptionDescriptor.getTitle(t, getAnyContext()), 1);
						return;
					}

					CrashHandler.showErrorDialog(new CrashHandler.CrashReport.Builder()
							.setTitle("Failed to check for updates")
							.setPrefix(R.string.please_report_bug_app)
							.setThrowable(t)
							.build());
				}
			});

			default -> toast("Unknown action: " + actionName);
		}
	}
}

File: app/src/main/java/com/mrboomdev/awery/ui/mobile/screens/settings/SettingsActions.kt
package com.mrboomdev.awery.ui.mobile.screens.settings

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings
import android.util.Log
import com.mrboomdev.awery.R
import com.mrboomdev.awery.app.App.Companion.showLoadingWindow
import com.mrboomdev.awery.app.App.Companion.toast
import com.mrboomdev.awery.app.AweryLifecycle.Companion.anyContext
import com.mrboomdev.awery.app.AweryLifecycle.Companion.appContext
import com.mrboomdev.awery.app.AweryLifecycle.Companion.getAnyActivity
import com.mrboomdev.awery.app.AweryLocales
import com.mrboomdev.awery.app.CrashHandler
import com.mrboomdev.awery.app.data.Constants
import com.mrboomdev.awery.app.data.Constants.DIRECTORY_IMAGE_CACHE
import com.mrboomdev.awery.app.data.settings.SettingsItem
import com.mrboomdev.awery.app.services.BackupService
import com.mrboomdev.awery.app.update.UpdatesManager
import com.mrboomdev.awery.app.update.UpdatesManager.showUpdateDialog
import com.mrboomdev.awery.generated.AwerySettings
import com.mrboomdev.awery.ui.mobile.screens.setup.SetupActivity
import com.mrboomdev.awery.util.ContentType
import com.mrboomdev.awery.util.exceptions.explain
import com.mrboomdev.awery.util.extensions.hasPermission
import com.mrboomdev.awery.util.extensions.requestPermission
import com.mrboomdev.awery.util.extensions.startActivity
import com.mrboomdev.awery.util.extensions.startActivityForResult
import com.mrboomdev.awery.util.extensions.startService
import com.mrboomdev.awery.util.ui.dialog.DialogBuilder
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.jetbrains.annotations.Contract
import xcrash.XCrash
import java.io.File
import java.util.Calendar
import kotlin.concurrent.thread
import kotlin.coroutines.cancellation.CancellationException

object SettingsActions {
	private const val TAG = "SettingsActions"

	@JvmStatic
	@Contract(pure = true)
	fun run(item: SettingsItem) {
		val actionName = item.key ?: return

		when(actionName) {
			AwerySettings.TRY_CRASH_NATIVE -> XCrash.testNativeCrash(false)
			AwerySettings.TRY_CRASH_JAVA -> XCrash.testJavaCrash(false)
			AwerySettings.TRY_CRASH_NATIVE_ASYNC -> thread { XCrash.testNativeCrash(false) }
			AwerySettings.TRY_CRASH_JAVA_ASYNC -> thread { XCrash.testJavaCrash(false) }

			AwerySettings.ABOUT -> anyContext.startActivity(AboutActivity::class)
			AwerySettings.START_ONBOARDING -> anyContext.startActivity(SetupActivity::class)
			AwerySettings.UI_LANGUAGE -> AweryLocales.showPicker(getAnyActivity()!!)

			AwerySettings.PLAYER_SYSTEM_SUBTITLES -> anyContext.startActivity(
				action = Settings.ACTION_CAPTIONING_SETTINGS)

			AwerySettings.SETUP_THEME -> anyContext.startActivity(
				SetupActivity::class, extras = mapOf(
					SetupActivity.EXTRA_STEP to SetupActivity.STEP_THEMING,
					SetupActivity.EXTRA_FINISH_ON_COMPLETE to true
				))

			AwerySettings.CLEAR_IMAGE_CACHE -> {
				File(anyContext.cacheDir, DIRECTORY_IMAGE_CACHE).deleteRecursively()
				toast(R.string.cleared_successfully)
			}

			AwerySettings.CLEAR_WEBVIEW_CACHE -> {
				File(anyContext.cacheDir, Constants.DIRECTORY_WEBVIEW_CACHE).deleteRecursively()
				toast(R.string.cleared_successfully)
			}

			AwerySettings.CLEAR_NET_CACHE -> {
				File(appContext.cacheDir, Constants.DIRECTORY_NET_CACHE).deleteRecursively()
				toast(R.string.cleared_successfully)
			}

			AwerySettings.BACKUP -> {
				val date = Calendar.getInstance()

				val defaultName = "awery_backup_[" + date[Calendar.YEAR] + "_" +
						date[Calendar.MONTH] + "_" +
						date[Calendar.DATE] + "]_[" +
						date[Calendar.HOUR_OF_DAY] + "_" +
						date[Calendar.MINUTE] + "].awerybck"

				val context = getAnyActivity<Activity>()!!

				if(Build.VERSION.SDK_INT <= Build.VERSION_CODES.P && !context.hasPermission(android.Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
					context.requestPermission(android.Manifest.permission.WRITE_EXTERNAL_STORAGE, { didGrant ->
						if(didGrant) run(item) else (DialogBuilder(context).apply {
							setTitle("Permission required!")
							setMessage("Sorry, but you cannot create files without an storage permission.")
							setNegativeButton(R.string.dismiss) { dismiss() }

							setPositiveButton("Open settings") {
								context.startActivity(action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
									data = Uri.parse("package:${context.packageName}"))

								dismiss()
							}
						}).show()
					})

					return
				}

				context.startActivityForResult(
					action = Intent.ACTION_CREATE_DOCUMENT,
					type = ContentType.ANY.mimeType,
					categories = arrayOf(Intent.CATEGORY_OPENABLE),
					extras = mapOf(Intent.EXTRA_TITLE to defaultName),
					callback = { resultCode, result ->
						if(resultCode != Activity.RESULT_OK) {
							return@startActivityForResult
						}

						context.startService(BackupService::class, BackupService.Args(
							BackupService.Action.BACKUP, result!!.data!!))
					})
			}

			AwerySettings.RESTORE -> {
				val context = getAnyActivity<Activity>()!!

				if(Build.VERSION.SDK_INT <= Build.VERSION_CODES.P && !context.hasPermission(android.Manifest.permission.READ_EXTERNAL_STORAGE)) {
					context.requestPermission(android.Manifest.permission.READ_EXTERNAL_STORAGE, { didGrant ->
						if(didGrant) run(item) else (DialogBuilder(context).apply {
							setTitle("Permission required!")
							setMessage("Sorry, but you cannot select files without an storage permission.")
							setNegativeButton(R.string.dismiss) { dismiss() }

							setPositiveButton("Open settings") {
								context.startActivity(action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
									data = Uri.parse("package:${context.packageName}"))

								dismiss()
							}
						}).show()
					})

					return
				}

				context.startActivityForResult(Intent.createChooser(Intent(Intent.ACTION_GET_CONTENT).apply {
					setType(ContentType.ANY.mimeType)
				}, "Choose a backup file"), { resultCode, result ->
					if(resultCode != Activity.RESULT_OK) return@startActivityForResult

					context.startService(BackupService::class, BackupService.Args(
						BackupService.Action.RESTORE, result!!.data!!))
				})
			}

			AwerySettings.CHECK_APP_UPDATE -> {
				val window = showLoadingWindow()

				CoroutineScope(Dispatchers.IO + CoroutineExceptionHandler { _, t ->
					Log.e(TAG, "Failed to check for updates!", t)
					window.dismiss()

					if(t is CancellationException) {
						toast(t.explain().title)
						return@CoroutineExceptionHandler
					}

					CrashHandler.showDialog(
						title = "Failed to check for updates",
						messagePrefixRes = R.string.please_report_bug_app,
						throwable = t)
				}).launch {
					val update = UpdatesManager.fetchLatestAppUpdate()
					showUpdateDialog(getAnyActivity<Activity>()!!, update)
					window.dismiss()
				}
			}

			else -> toast("Unknown action: $actionName")
		}
	}
}